metrics/exporter.go

```diff
@@ -32,11 +33,99 @@ type flushable interface {
 	Flush()
 }
 
+// ExporterOptions contains options for configuring the exporter.
+type ExporterOptions struct {
+	// Domain is the metrics domain. e.g. "knative.dev". Must be present.
+	//
+	// Stackdriver uses the following format to construct full metric name:
+	//    <domain>/<component>/<metric name from View>
+	// Prometheus uses the following format to construct full metric name:
+	//    <component>_<metric name from View>
+	// Domain is actually not used if metrics backend is Prometheus.
+	Domain string
+
+	// Component is the name of the component that emits the metrics. e.g.
+	// "activator", "queue_proxy". Should only contains alphabets and underscore.
+	// Must be present.
+	Component string
+
+	// PrometheusPort is the port to expose metrics if metrics backend is Prometheus.
+	// It should be between maxPrometheusPort and maxPrometheusPort. 0 value means
+	// using the default 9090 value. If is ignored if metrics backend is not
+	// Prometheus.
+	PrometheusPort int
+
+	// ConfigMap is the data from config map config-observability. Must be present.
+	// See https://github.com/knative/serving/blob/master/config/config-observability.yaml
+	// for details.
+	ConfigMap map[string]string
+}
+
+// UpdateExporterFromConfigMap returns a helper func that can be used to update the exporter
+// when a config map is updated.
+func UpdateExporterFromConfigMap(component string, logger *zap.SugaredLogger) func(configMap *corev1.ConfigMap) {
+	domain := Domain()
+	return func(configMap *corev1.ConfigMap) {
+		UpdateExporter(ExporterOptions{
+			Domain:    domain,
+			Component: component,
+			ConfigMap: configMap.Data,
+		}, logger)
+	}
+}
+
+// UpdateExporter updates the exporter based on the given ExporterOptions.
+func UpdateExporter(ops ExporterOptions, logger *zap.SugaredLogger) error {
+	metricsMux.Lock()
```

Comment:
Could you add the comment here to explain why we lock the whole function?

Reply:
Pending